In [1]:
import os
os.chdir("D://Master//10. TFM//Data")

In [2]:
import pandas as pd
import numpy as np

In [3]:
df_inicial = pd.read_csv("df_inicial.csv")
df_inicial.head()

customer_id  customer_zip_code_prefix  \
0  3ce436f183e68e07877b285a838db11a                     28013   
1  e6eecc5a77de221464d1c4eaff0a9b64                     75800   
2  4ef55bf80f711b372afebcb7c715344a                     30720   
3  30407a72ad8b3f4df4d15369126b20c9                     83070   
4  91a792fef70ecd8cc69d3c7feb3d12da                     36400   

           customer_city customer_state                          order_id  \
0  campos dos goytacazes             RJ  00010242fe8c5a6d1ba2dd792cb16214   
1                  jatai             GO  130898c0987d1801452a8ed92a670612   
2         belo horizonte             MG  532ed5e14e24ae1f0d735b91524b98b9   
3   sao jose dos pinhais             PR  6f8c31653edb8c83e1a739408b5ff750   
4   conselheiro lafaiete             MG  7d19f4ef4d04461989632411b7e588b9   

   price  freight_value payment_type  payment_value  \
0   58.9          13.29  credit_card          72.19   
1   55.9          17.96       boleto          73.86   
2   64.9          18.33  credit_card          83.23   
3   58.9          16.17  credit_card          75.07   
4   58.9          13.29  credit_card          72.19   

                         product_id product_category_name order_status  \
0  4244733e06e7ecb4970a6e2683c13e61            cool_stuff    delivered   
1  4244733e06e7ecb4970a6e2683c13e61            cool_stuff    delivered   
2  4244733e06e7ecb4970a6e2683c13e61            cool_stuff    delivered   
3  4244733e06e7ecb4970a6e2683c13e61            cool_stuff    delivered   
4  4244733e06e7ecb4970a6e2683c13e61            cool_stuff    delivered   

  order_purchase_timestamp    order_approved_at  \
0      2017-09-13 08:59:02  2017-09-13 09:45:35   
1      2017-06-28 11:52:20  2017-06-29 02:44:11   
2      2018-05-18 10:25:53  2018-05-18 12:31:43   
3      2017-08-01 18:38:42  2017-08-01 18:55:08   
4      2017-08-10 21:48:40  2017-08-10 22:05:11   

                          review_id  review_score  
0  97ca439bc427b48bc1cd7177abe71365             5  
1  b11cba360bbe71410c291b764753d37f             5  
2  af01c4017c5ab46df6cc810e069e654a             4  
3  8304ff37d8b16b57086fa283fe0c44f8             5  
4  426f43a82185969503fb3c86241a9535             5

In [4]:
df_inicial.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112369 entries, 0 to 112368
Data columns (total 16 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   customer_id               112369 non-null  object 
 1   customer_zip_code_prefix  112369 non-null  int64  
 2   customer_city             112369 non-null  object 
 3   customer_state            112369 non-null  object 
 4   order_id                  112369 non-null  object 
 5   price                     112369 non-null  float64
 6   freight_value             112369 non-null  float64
 7   payment_type              112369 non-null  object 
 8   payment_value             112369 non-null  float64
 9   product_id                112369 non-null  object 
 10  product_category_name     110771 non-null  object 
 11  order_status              112369 non-null  object 
 12  order_purchase_timestamp  112369 non-null  object 
 13  order_approved_at         112354 non-null  o

Primero observamos que necesitamos modificar el tipo de variable en algunos casos:
- customer_zip_code_prefix debe ser categórica aunque sea un número
- customer_city debe ser categórica
- customer_state debe ser categórica
- product_category_name debe ser categórica
- order_status debe ser categórica
- order_purchase_timestamp debería dividirse en año, mes, día, hora y minuto.
- review_score debería ser categórica

In [5]:
df_inicial['customer_zip_code_prefix'] = df_inicial['customer_zip_code_prefix'].astype('category')
df_inicial['customer_city'] = df_inicial['customer_city'].astype('category')
df_inicial['customer_state'] = df_inicial['customer_state'].astype('category')
df_inicial['product_category_name'] = df_inicial['product_category_name'].astype('category')
df_inicial['order_status'] = df_inicial['order_status'].astype('category')
df_inicial['review_score'] = df_inicial['review_score'].astype('category')

Convertimos order_purchase_timestamp en formato datetime para poder separarlo y convertirlo en categórica

In [6]:
df_inicial['order_purchase_timestamp'] = pd.to_datetime(df_inicial['order_purchase_timestamp'])

In [7]:
df_inicial['year'] = df_inicial['order_purchase_timestamp'].dt.year
df_inicial['month'] = df_inicial['order_purchase_timestamp'].dt.month
df_inicial['day_of_week'] = df_inicial['order_purchase_timestamp'].dt.dayofweek
df_inicial['hour'] = df_inicial['order_purchase_timestamp'].dt.hour

In [8]:
df_inicial['year'] = df_inicial['year'].astype('category')
df_inicial['month'] = df_inicial['month'].astype('category')
df_inicial['day_of_week'] = df_inicial['day_of_week'].astype('category')
df_inicial['hour'] = df_inicial['hour'].astype('category')

In [9]:
df_inicial.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112369 entries, 0 to 112368
Data columns (total 20 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   customer_id               112369 non-null  object        
 1   customer_zip_code_prefix  112369 non-null  category      
 2   customer_city             112369 non-null  category      
 3   customer_state            112369 non-null  category      
 4   order_id                  112369 non-null  object        
 5   price                     112369 non-null  float64       
 6   freight_value             112369 non-null  float64       
 7   payment_type              112369 non-null  object        
 8   payment_value             112369 non-null  float64       
 9   product_id                112369 non-null  object        
 10  product_category_name     110771 non-null  category      
 11  order_status              112369 non-null  category      
 12  or

Ahora que ya tenemos el formato correcto, vamos a limpiar los valores nulos y observar los únicos

In [10]:
valores_nulos = df_inicial.isnull().sum()
valores_nulos

customer_id                    0
customer_zip_code_prefix       0
customer_city                  0
customer_state                 0
order_id                       0
price                          0
freight_value                  0
payment_type                   0
payment_value                  0
product_id                     0
product_category_name       1598
order_status                   0
order_purchase_timestamp       0
order_approved_at             15
review_id                      0
review_score                   0
year                           0
month                          0
day_of_week                    0
hour                           0
dtype: int64

In [11]:
valores_unicos = df_inicial.nunique()
valores_unicos

customer_id                 97916
customer_zip_code_prefix    14955
customer_city                4108
customer_state                 27
order_id                    97916
price                        5948
freight_value                6976
payment_type                    5
payment_value               27474
product_id                  32789
product_category_name          73
order_status                    7
order_purchase_timestamp    97370
order_approved_at           89533
review_id                   97708
review_score                    5
year                            3
month                          12
day_of_week                     7
hour                           24
dtype: int64

Eliminamos las filas con algún valor único

In [12]:
df_inicial = df_inicial.dropna()

In [13]:
df_inicial.isnull().sum()

customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
order_id                    0
price                       0
freight_value               0
payment_type                0
payment_value               0
product_id                  0
product_category_name       0
order_status                0
order_purchase_timestamp    0
order_approved_at           0
review_id                   0
review_score                0
year                        0
month                       0
day_of_week                 0
hour                        0
dtype: int64

In [14]:
df_inicial.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 110757 entries, 0 to 112368
Data columns (total 20 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   customer_id               110757 non-null  object        
 1   customer_zip_code_prefix  110757 non-null  category      
 2   customer_city             110757 non-null  category      
 3   customer_state            110757 non-null  category      
 4   order_id                  110757 non-null  object        
 5   price                     110757 non-null  float64       
 6   freight_value             110757 non-null  float64       
 7   payment_type              110757 non-null  object        
 8   payment_value             110757 non-null  float64       
 9   product_id                110757 non-null  object        
 10  product_category_name     110757 non-null  category      
 11  order_status              110757 non-null  category      
 12  or

Observamos que se han eliminado todos los valores nulos a costa de reducir el número de entradas también.

In [15]:
df_inicial

customer_id customer_zip_code_prefix  \
0       3ce436f183e68e07877b285a838db11a                    28013   
1       e6eecc5a77de221464d1c4eaff0a9b64                    75800   
2       4ef55bf80f711b372afebcb7c715344a                    30720   
3       30407a72ad8b3f4df4d15369126b20c9                    83070   
4       91a792fef70ecd8cc69d3c7feb3d12da                    36400   
...                                  ...                      ...   
112364  050309b91cc5e04e68841938e7984aaf                    28740   
112365  11e0f43ab4e2d2c48348dd9332c0ef80                    83065   
112366  dec8952e97ef6124259c56914fb3569c                    27261   
112367  a5201e1a6d71a8d21e869151bd5b4085                    59955   
112368  96d649da0cc4ff33bb408b199d4c7dcf                    18605   

                customer_city customer_state  \
0       campos dos goytacazes             RJ   
1                       jatai             GO   
2              belo horizonte             MG   
3        sao jose dos pinhais             PR   
4        conselheiro lafaiete             MG   
...                       ...            ...   
112364    conceicao de macabu             RJ   
112365   sao jose dos pinhais             PR   
112366          volta redonda             RJ   
112367        tenente ananias             RN   
112368               botucatu             SP   

                                order_id  price  freight_value payment_type  \
0       00010242fe8c5a6d1ba2dd792cb16214   58.9          13.29  credit_card   
1       130898c0987d1801452a8ed92a670612   55.9          17.96       boleto   
2       532ed5e14e24ae1f0d735b91524b98b9   64.9          18.33  credit_card   
3       6f8c31653edb8c83e1a739408b5ff750   58.9          16.17  credit_card   
4       7d19f4ef4d04461989632411b7e588b9   58.9          13.29  credit_card   
...                                  ...    ...            ...          ...   
112364  ffebd80e3291e811c308365936897efd  129.9          51.20  credit_card   
112365  ffee31fb4b5e35c9123608015637c495   99.0          13.52  credit_card   
112366  fff7c4452f050315db1b3f24d9df5fcd  736.0          20.91  credit_card   
112367  fffa82886406ccf10c7b4e35c4ff2788  229.9          44.02  credit_card   
112368  fffe41c64501cc87c801fd61db3f6244   43.0          12.79  credit_card   

        payment_value                        product_id product_category_name  \
0               72.19  4244733e06e7ecb4970a6e2683c13e61            cool_stuff   
1               73.86  4244733e06e7ecb4970a6e2683c13e61            cool_stuff   
2               83.23  4244733e06e7ecb4970a6e2683c13e61            cool_stuff   
3               75.07  4244733e06e7ecb4970a6e2683c13e61            cool_stuff   
4               72.19  4244733e06e7ecb4970a6e2683c13e61            cool_stuff   
...               ...                               ...                   ...   
112364         181.10  4cc4d02efc8f249c13355147fb44e34d    ferramentas_jardim   
112365         112.52  b10ecf8e33aaaea419a9fa860ea80fb5      moveis_decoracao   
112366         756.91  dd469c03ad67e201bc2179ef077dcd48    relogios_presentes   
112367         273.92  bbe7651fef80287a816ead73f065fc4b         esporte_lazer   
112368          55.79  350688d9dc1e75ff97be326363655e01       cama_mesa_banho   

       order_status order_purchase_timestamp    order_approved_at  \
0         delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1         delivered      2017-06-28 11:52:20  2017-06-29 02:44:11   
2         delivered      2018-05-18 10:25:53  2018-05-18 12:31:43   
3         delivered      2017-08-01 18:38:42  2017-08-01 18:55:08   
4         delivered      2017-08-10 21:48:40  2017-08-10 22:05:11   
...             ...                      ...                  ...   
112364    delivered      2018-07-13 20:04:05  2018-07-13 20:23:55   
112365    delivered      2018-08-18 10:00:59  2018-08-18 10:10:11   
112366    delivered      2017-06-01 16:53:03  2017-06-01 17:05:23   
112367    d

El dataset que obtenemos tiene 110757 entradas. Cada una de ellas representa un pedido realizado en la tienda online y nos proporciona información relevante al cliente que lo ha realizado.

Hacemos el análisis de cada columna:

- Customer_id: Esta columna es un valor clave que identifica al cliente. No nos aporta información por lo que habrá que eliminarla de cara a los análisis posteriores.
- Customer_zip_code_prefix: Variable categórica que representa el código postal del cliente. Hay 14955 valores únicos por lo que la información es muy dispersa.
- Customer city: Variable categórica con 4108 ciudades diferentes. Habrá que utilizar 'one-hot-encoding' para asignar una variable numérica binaria a cada ciudad dado que hay algoritmos que sólo interpretan variables numéricas. Esto tiene como consecuencia que se crearán multitud de variables adicionales y no todas resultarán de utilidad. Tenemos como opción eliminar la variable ciudad y utilizar en su lugar la de estado, que tiene sólo 27 valores.
- Customer_state: Variable categórica con 27 estados diferentes. Podemos usar label-encoding o one-hot-encoding y ver la diferencia entre uno u otro. Al no existir superioridad entre los diferentes valores, tal vez sea mejor utilizar one-hot-encoding.
- Order_id: Al igual que customer_id esta columna es un valor clave que identifica a cada uno de los pedidos. No nos aporta información por lo que habrá que eliminarla de cara a los análisis posteriores.
- Price: Variable numérica que representa el precio del pedido sin incluir el porte.
- Freight_value: Variable numérica que representa el precio del porte de cada pedido.
- Payment_type: Variable categórica que representa el tipo de método de pago. Será necesario realizar label-encoding para realizar los análisis.
- Payment_value: Se trata de una variable numérica que representa el valor total del pedido, siendo la suma del valor del artículo y del porte.
- Product_id: Al igual que customer_id esta columna es un valor clave que identifica a cada uno de los productos. No nos aporta información por lo que habrá que eliminarla de cara a los análisis posteriores.
- Product_category_name: Variable categórica que representa a los 73 tipos de producto que se pueden comprar en la tienda online. Será necesario convertirla en numérica mediante label-encoding / one-hot-encoding
- Order_status: Variable categórica que representa los 7 estados en los que se puede encontrar un pedido. Habrá que realizar label-encoding ya que sí existe cierto orden o jerarquía entre ellos.
- Year (variable categórica del año)
- Month (variable categórica del mes)
- Day_of_week (variable categórica del día de la semana)
- Hour (variable categórica de la hora)
- Order_approved_at: Esta variable representa la hora en que se aprovó el pedido. Puede no tener exceso valor de cara al análisis por lo que podríamos eliminarla.
- Review_id: Esta columna es un valor clave que identifica a la valoración del pedido. No nos aporta información por lo que habrá que eliminarla de cara a los análisis posteriores.
- Review_score: Esta variable categórica representa la nota que ha puntuado el cliente en cada pedido.

A continuación, vamos a eliminar las variables identificativas (ID) que no aportan valor en sí para nuestro estudio, además de la variable 'order_approved_at'.

In [16]:
df = df_inicial[['customer_zip_code_prefix', 'customer_city', 'customer_state', 'price', 'freight_value', 'payment_type','payment_value', 'product_category_name', 'order_status', 'review_score', 'year', 'month', 'day_of_week', 'hour']]

df.head()

customer_zip_code_prefix          customer_city customer_state  price  \
0                    28013  campos dos goytacazes             RJ   58.9   
1                    75800                  jatai             GO   55.9   
2                    30720         belo horizonte             MG   64.9   
3                    83070   sao jose dos pinhais             PR   58.9   
4                    36400   conselheiro lafaiete             MG   58.9   

   freight_value payment_type  payment_value product_category_name  \
0          13.29  credit_card          72.19            cool_stuff   
1          17.96       boleto          73.86            cool_stuff   
2          18.33  credit_card          83.23            cool_stuff   
3          16.17  credit_card          75.07            cool_stuff   
4          13.29  credit_card          72.19            cool_stuff   

  order_status review_score  year month day_of_week hour  
0    delivered            5  2017     9           2    8  
1    delivered            5  2017     6           2   11  
2    delivered            4  2018     5           4   10  
3    delivered            5  2017     8           1   18  
4    delivered            5  2017     8           3   21

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 110757 entries, 0 to 112368
Data columns (total 14 columns):
 #   Column                    Non-Null Count   Dtype   
---  ------                    --------------   -----   
 0   customer_zip_code_prefix  110757 non-null  category
 1   customer_city             110757 non-null  category
 2   customer_state            110757 non-null  category
 3   price                     110757 non-null  float64 
 4   freight_value             110757 non-null  float64 
 5   payment_type              110757 non-null  object  
 6   payment_value             110757 non-null  float64 
 7   product_category_name     110757 non-null  category
 8   order_status              110757 non-null  category
 9   review_score              110757 non-null  category
 10  year                      110757 non-null  category
 11  month                     110757 non-null  category
 12  day_of_week               110757 non-null  category
 13  hour                      110

In [18]:
df.nunique()

customer_zip_code_prefix    14907
customer_city                4093
customer_state                 27
price                        5882
freight_value                6954
payment_type                    5
payment_value               27306
product_category_name          73
order_status                    7
review_score                    5
year                            3
month                          12
day_of_week                     7
hour                           24
dtype: int64

In [19]:
df['payment_type'].value_counts()


credit_card    82205
boleto         22503
varios          3215
debit_card      1661
voucher         1173
Name: payment_type, dtype: int64

In [21]:
df

customer_zip_code_prefix          customer_city customer_state  price  \
0                         28013  campos dos goytacazes             RJ   58.9   
1                         75800                  jatai             GO   55.9   
2                         30720         belo horizonte             MG   64.9   
3                         83070   sao jose dos pinhais             PR   58.9   
4                         36400   conselheiro lafaiete             MG   58.9   
...                         ...                    ...            ...    ...   
112364                    28740    conceicao de macabu             RJ  129.9   
112365                    83065   sao jose dos pinhais             PR   99.0   
112366                    27261          volta redonda             RJ  736.0   
112367                    59955        tenente ananias             RN  229.9   
112368                    18605               botucatu             SP   43.0   

        freight_value payment_type  payment_value product_category_name  \
0               13.29  credit_card          72.19            cool_stuff   
1               17.96       boleto          73.86            cool_stuff   
2               18.33  credit_card          83.23            cool_stuff   
3               16.17  credit_card          75.07            cool_stuff   
4               13.29  credit_card          72.19            cool_stuff   
...               ...          ...            ...                   ...   
112364          51.20  credit_card         181.10    ferramentas_jardim   
112365          13.52  credit_card         112.52      moveis_decoracao   
112366          20.91  credit_card         756.91    relogios_presentes   
112367          44.02  credit_card         273.92         esporte_lazer   
112368          12.79  credit_card          55.79       cama_mesa_banho   

       order_status review_score  year month day_of_week hour  
0         delivered            5  2017     9           2    8  
1         delivered            5  2017     6           2   11  
2         delivered            4  2018     5           4   10  
3         delivered            5  2017     8           1   18  
4         delivered            5  2017     8           3   21  
...             ...          ...   ...   ...         ...  ...  
112364    delivered            5  2018     7           4   20  
112365    delivered            4  2018     8           5   10  
112366    delivered            5  2017     6           3   16  
112367    delivered            4  2017    12           0   16  
112368    delivered            5  2018     6           5   17  

[110757 rows x 14 columns]

Guardamos este dataframe en un archivo csv

In [20]:
df.to_csv("df.csv")